src/System.CommandLine.StaticCompletions/CompletionsCommand.cs

```diff
@@ -0,0 +1,105 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.CommandLine.StaticCompletions;
+
+using System.CommandLine;
+using System.CommandLine.StaticCompletions.Shells;
+using System.Linq;
+
+public class CompletionsCommand : CliCommand
+{
+    public static IShellProvider[] DefaultShells = [
+        new BashShellProvider(),
+        new PowershellShellProvider(),
+        new FishShellProvider(),
+        new ZshShellProvider(),
+        new NushellShellProvider()
+    ];
+
+    /// <summary>
+    /// Creates a new Completions command exclusively with the provided supported shells. If no shells are provided, the <see cref="DefaultShells"/> are used.
+    /// </summary>
+    /// <param name="supportedShells">The shells to support in the completions command. If null, <see cref="DefaultShells"/> will be used.</param>
+    public CompletionsCommand(IEnumerable<IShellProvider>? supportedShells = null) : this((supportedShells ?? DefaultShells).ToDictionary(s => s.ArgumentName, StringComparer.OrdinalIgnoreCase))
+    { }
+
+    private CompletionsCommand(Dictionary<string, IShellProvider> shellMap)
+        : base("completions", "Commands for generating and registering completions for supported shells")
+    {
+        var shellArg = new CliArgument<IShellProvider>("shell")
+        {
+            Description = "The shell for which to generate or register completions",
+            Arity = ArgumentArity.ZeroOrOne,
+            // called when no token is presented at all
+            DefaultValueFactory = (argResult) => LookupShellFromEnvironment(shellMap),
+            // called for all other scenarios
+            CustomParser = (argResult) =>
+            {
+                return argResult.Tokens switch
+                {
+                // shouldn't be required because of the DefaultValueFactory above
+                [] => LookupShellFromEnvironment(shellMap),
+                [var shellToken] => shellMap[shellToken.Value],
+                    _ => throw new InvalidOperationException("Unexpected number of tokens")
+                };
+            }
+        };
+
+        shellArg.AcceptOnlyFromAmong(shellMap.Keys.ToArray());
+        Subcommands.Add(new GenerateScriptCommand(shellArg));
+        Subcommands.Add(new RegisterScriptCommand(shellArg));
+
+        static IShellProvider LookupShellFromEnvironment(Dictionary<string, IShellProvider> shellMap)
+        {
+            if (OperatingSystem.IsWindows())
+            {
+                return shellMap[PowershellShellProvider.PowerShell];
```

Comment:
I don't normally use pwsh on windows...

Reply:
I'm guessing you use `cmd`? First off, it's 2024 (almost 2025 dude 😛). Secondly, `cmd` has no good completions story anyway so there's no sense pretending it even exists for this feature. Finally, this is just the default value computation here - I'd expect users that worry about drift here to add an actual value for the shell a la `dotnet completions generate pwsh` in their $PROFILE or equivalent.

Reply:
Hey, some people use vim...

I'm not really worried about the lack of completions; I just want to make sure that we don't make assumptions that end up causing issues for people using unexpected shells. I didn't test what would happen if I tried to use this on cmd, so if it just doesn't do anything, that's fine, but if it tries to show completions and explodes, that's not fine.

When I looked at your demos in the description, it seemed like you had to explicitly run the completions.ps1 script to get completions. Is that the only way to ever get completions? I kinda thought there was like a tab-complete option or something, and if so, I'd just want that to not break current cmd users.